### DISCUSSION TITLE
new lifecycle `onAfterUnmount`

### DISCUSSION THREAD

**Discussion by Alfred-Skyblue** [vuejs/rfcs#510] [RFC Discussions]
When we remove an element using `el.remove()`, the components contained within that element do not execute the `onUnmounted` lifecycle. I would like to add a new lifecycle called `onAfterUnmounted` that triggers after the component's `el` or the element containing the current component is removed from the DOM tree. Otherwise, when we manually remove DOM elements, it causes memory leaks in the child components contained within that element.

[demo](https://play.vuejs.org/#eNqVVtty00gQ/ZVGtYXl4JJx8WZkFVnI1rKFyRaXp4gHWRo7ItKMajQyDsH/TvdcdLEVKOwHjbpP9/TpOdP2g3dZVcG+Yd7SC+tU5pWCmqmmgiLhu1XsqTr2opjnZSWkggeQbAtH2EpRwgTDJq3rtSgraw/m9EJZh+7FwL+wgFTwWkFK4StKH/77af3uqmAl4yryp33A4hFEzMO5KR5LxRfFyqpIFMM3gHDTKCU4vEqLPL1DSpQpkKwUe4bczELXF84N9PGwxWjcYhiY5Xuq0oYgNCRQFGrW+ED/GHDhkAsLpafBhvOWkTfDI8GObPNd8LUWHM/tgZKZHHnB5HWlcuxY7C1Be8iXFIX49p+2KdmwmbOntyy9G7F/rQ9ki73/JauZJMKtTyVyx5RxX318zw64bp2lyJoC0b9wfmC1KBqq0cD+bniGZfdwutq3Wjc5332qrw6K8dqRokIJedT42EMVUbceo96V+yJ4oeNifsQuOomeCr8vdsE/81I0XLHsRPT07Xl9fwqryG6aYCXKn+gL0cNMUKjH34nVqCIyahw7/Pbm/Lruy61i8rz4YH7qmrx0ZIb2AaM+pcVZ9j8hZpb00bmsZ5Tp6TaBqpFzyxEP9XUjJY6AtzgeEp6yGVa2HhLW6og9y5EddOy24Snp5IyIn26W/nQV7UWeTS1zM3z2JU6e8x39aVBJcbg30HZ37N0qsvHasdF36I0oP9DQSAp/XwZ/sWKWbrBthKD2oS7pazY0ql0nNBM5+wa4Ct+LDDl+ZCr8xzKI7PQrmHKbyCWsG5WQ99pa4AfwpigoEz4IP7+4wH0vwCLgzfUaB5EuDcQWErs9MDNjNfZVlciktJ6+JcXruknSO7LNR5p8Rp4omzRL0JzI4LIswZHDst3y5kvMp0aLRDXdYBOQTtukAB++edP9AMi34D/ROBuHO9goaieuqHNk7pLgNbJJZgasEXgsJt2llMl9kNf66bt6p11+awm2Ql4l6S2qiW6QzvQkSDKS19SkxMbWbFCX8bucdl+3szvaKR6SaqS2O5vTx8mZ+6U11O9ylFN3kylfj3L+ncFqtYLn/dwASAF8o0SXiIQxSNoSd5fE/CJmdKQ1luXA9ifT2F0EldEC1H1lqpikt3mRUfIJPH06SBgUjO/ULURUau9u6UpJFDnP2AG3ff7SLsOxeOt89myQZISB/qfRhd/oMJRhF9Lr0c1ANl+oVW2PTzYy1HupA0yhkpzXTsBnAU4ip7ryjZr6n+5kM7y7inWXYojrn7X7aKGPvLRLu9APGlidCAO78DORNjQxZjig27O0/y2gbjZKMmZeu5HnHX8CpO6Aow==)

In the demo above, we demonstrate the usage scenario where `onUnmounted` is not triggered when we use `el.remove()` to delete elements.

[PR](https://github.com/vuejs/core/pull/8762)

**Comment by Justineo** [comment]
I believe we shouldn’t manually remove elements that are created by Vue. What is the use case for this?

**Reply by Alfred-Skyblue** [reply]
For example, in a back-office management system, I am using imperative rendering for a dialog component. Since it is not rendered within the template, the dialog does not trigger the `unmounted` event when the current component instance is destroyed. The dialog component covers the entire screen and is mounted to the `body` element. However, due to certain reasons such as a change in permissions or an expired token, I need to redirect the route to the login page. During this redirection process, I need to manually remove the DOM element that contains the dialog. Many UI libraries, such as [Element Plus MessageBox](https://element-plus.org/en-US/component/message-box.html) and[ Ant Design Vue Modal](https://www.antdv.com/components/modal), utilize imperative rendering for components.

**Reply by Justineo** [reply]
This is rather a matter of API design of component libraries I think. The real problem is that these imperatively triggered dialogs do not provide a way to safely dispose themselves.

On the other hand, using `MutationObserver` to observe the *output* of the rendering logic looks very unsafe to me. DOM elements can be removed from its parent element for various reasons, ie. due to `v-if` switches, portals, or DOM manipulation performed by users or third party libraries.

If Vue was adding a `onAfterUnmount` hook, it should be when Vue itself has done removing a component's node from the DOM (rather than it found out from the output that the node is gone). This has a more legit use case IMO. eg. when a component is `v-if`ed out but inside a `<Transition>`, we may want to dispose its contents after transition completes (disposing third party chart instance, rich-text editors, etc.), otherwise we will see the content disappears immediately after the component starts to fade/slide out.
